### PR TITLE
#33 #34 Added support for loading certificates from files and base64 …

### LINKED ISSUE #34: Add support for other certificate sources than the certificate store
Loading certificates from other sources than the certificate store, which might be hard to access on cloud agents, should be supported.

### PR DESCRIPTION
**Pull request by EkiH**
…strings. Added support for loading multiple certificates to the request so the full chain can be provided.

### CODE REVIEW

File: Frends.Web/Web.cs [hunk 1]
```diff
@@ -126,11 +125,46 @@ public class Options : IEquatable<Options>
...
+        /// </summary>
+        [UIHint(nameof(Frends.Web.Authentication), "", Authentication.ClientCertificate)]
+        [DefaultValue(CertificateSource.CertificateStore)]
+        public CertificateSource ClientCertificateSource { get; set; }
+
+        /// <summary>
+        /// Path to the Client Certificate when using a file as the Certificate Source, pfx (pkcs12) files are recommended. For other supported formats, see
+        /// https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.import?view=netframework-4.7.1
+        /// </summary>
+        [UIHint(nameof(Frends.Web.Authentication), "", Authentication.ClientCertificate)]
```
Comment (nickeeex) [inline Frends.Web/Web.cs:139]: Is this supposed to be checking on the CertificateSource instead?  Could we hide the extra inputs that way? Or does the editor only support one layer of UIHints?

File: Frends.Web/Web.cs [hunk 2]
```diff
@@ -173,6 +209,9 @@ public bool Equals(Options other)
                    string.Equals(Password, other.Password) &&
                    string.Equals(Token, other.Token) &&
                    string.Equals(CertificateThumbprint, other.CertificateThumbprint) &&
+                   ClientCertificateSource == other.ClientCertificateSource &&
+                   ClientCertificateInBase64 == other.ClientCertificateInBase64 &&
+                   ClientCertificateFilePath == other.ClientCertificateFilePath &&
```
Comment (nickeeex) [inline Frends.Web/Web.cs:214]: LoadEntireChainForCertificate missing from equality check

File: Frends.Web/Web.cs [hunk 3]
```diff
@@ -651,7 +700,59 @@ internal static void SetDefaultRequestHeadersBasedOnOptions(this HttpClient http
...
+        private static X509Certificate2[] GetCertificatesFromFile(string clientCertificateFilePath, string keyPhrase)
+        {
+            return LoadCertificatesFromBytes(File.ReadAllBytes(clientCertificateFilePath), keyPhrase);
+        }
+
+        private static X509Certificate2[] GetCertificatesFromStore(string thumbprint,
+            bool loadEntireChain)
         {
             thumbprint = Regex.Replace(thumbprint, @"[^\da-zA-z]", string.Empty).ToUpper();
             var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
```
Comment (nickeeex) [inline Frends.Web/Web.cs]: When we are modifying this, is there ever a need to parameterize the store location? If someone had the client certificate in the local machine?

Comment (nickeeex) [inline Frends.Web/Web.cs]: The X509Store is disposable from .net framework 4.6 => 
This could be changed to using instead of finally close() ?

**Review by nickeeex — COMMENTED** [review]
Do the tests pass with these changed? There are some certificate tests.

File: Frends.Web/Web.cs (hunk 1 above)
Comment (EkiH) [reply Frends.Web/Web.cs:139]: Yes, sadly the editor/task loader supports only one UIHint per property. So if we used CertificateSource, it the certificate options would be shown when ClientCertificate authentication is not used. I thought it is clearer this way as client certificates are not used that often

File: Frends.Web/Web.cs (hunk 3 above)
Comment (EkiH) [reply Frends.Web/Web.cs]: I think the user store is the correct one and can't really imagine a case where the machine store would be used for client certificates

**Pull request merged**